core/src/main/scala/cats/derived/contravariant.scala

```diff
@@ -0,0 +1,98 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cats
+package derived
+
+import shapeless._
+import util.VersionSpecific.OrElse
+
+import scala.annotation.implicitNotFound
+
+@implicitNotFound("Could not derive an instance of Contravariant[${F}]")
+trait MkContravariant[F[_]] extends Contravariant[F] {
+  def safeContramap[A, B](fa: F[A])(f: B => Eval[A]): Eval[F[B]]
+  def contramap[A, B](fa: F[A])(f: B => A): F[B] = safeContramap(fa)(f andThen Eval.now).value
+}
+object MkContravariant extends MkContravariantDerivation {
+  def apply[F[_]](implicit F: MkContravariant[F]): MkContravariant[F] = F
+}
+private[derived] abstract class MkContravariantDerivation extends MkContravariantNested1 {
+  implicit val mkContraHNil: MkContravariant[Const[HNil]#λ] = mkContraConst
+  implicit val mkContraCNil: MkContravariant[Const[CNil]#λ] = mkContraConst
+
+  implicit def mkContraConst[T]: MkContravariant[Const[T]#λ] =
+    new MkContravariant[Const[T]#λ] {
+      def safeContramap[A, B](t: T)(f: B => Eval[A]): Eval[T] = Eval.now(t)
+    }
+}
+
+private[derived] abstract class MkContravariantNested1 extends MkContravariantCons {
+  implicit def mkFunctorContravariantNested[F[_]](implicit F: Split1[F, Functor, ContraOrMk]): MkContravariant[F] =
+    new MkContravariant[F] {
+      def safeContramap[A, B](fa: F[A])(f: B => Eval[A]): Eval[F[B]] =
+        Eval.later(F.fo.map(F.unpack(fa))(
+          (a: F.I[A]) => mkContraSafe(F.fi)(a)(f).value
+        )).map(F.pack[B])
+    }
+}
+
+private[derived] abstract class MkContravariantCons extends MkContravariantNested0 {
+  implicit def mkContraHCons[F[_]](implicit F: IsHCons1[F, ContraOrMk, MkContravariant]): MkContravariant[F] =
+    new MkContravariant[F] {
+      def safeContramap[A, B](fa: F[A])(f: B => Eval[A]): Eval[F[B]] =
+        Eval.now(F.unpack(fa)).flatMap { case (fha, fta) =>
+          for {
+            fhb <- mkContraSafe(F.fh)(fha)(f)
+            ftb <- F.ft.safeContramap(fta)(f)
+          } yield F.pack(fhb, ftb)
+        }
+    }
+
+  implicit def mkContraCCons[F[_]](implicit F: IsCCons1[F, ContraOrMk, MkContravariant]): MkContravariant[F] =
+    new MkContravariant[F] {
+      def safeContramap[A, B](fa: F[A])(f: B => Eval[A]): Eval[F[B]] = F.unpack(fa) match {
+        case Left(fha) => mkContraSafe(F.fh)(fha)(f).map(fhb => F.pack(Left(fhb)))
+        case Right(fta) => F.ft.safeContramap(fta)(f).map(ftb => F.pack(Right(ftb)))
+      }
+    }
+}
+private[derived] abstract class MkContravariantNested0 extends MkContravariantGeneric {
+  implicit def mkContravariantFunctorNested[F[_]](implicit F: Split1[F, ContraOrMk, Functor]): MkContravariant[F] =
+    new MkContravariant[F] {
+      def safeContramap[A, B](fa: F[A])(f: B => Eval[A]): Eval[F[B]] =
+        mkContraSafe(F.fo)(F.unpack(fa))(
+          (a: F.I[B]) => Eval.later(F.fi.map(a)(f andThen (_.value)))
+        ).map(F.pack[B])
+    }
```

Comment:
Interestingly, if this instance is higher in the chain, it sends the compiler into an infinite loop during autoresolution tests, which is kinda interesting.

Reply:
Hmm, does it have anything to do with Id?

Reply:
Possibly? I end up seeing a loop in the implicit log around various Generic10, Generic11, and Generic13 instances IIRC? I'm gonna mess around with 2.12/2.11 a bit more and see if I can get it to work on there.

Reply:
Sorry that I haven't found time to help you, I'm quite busy these two weeks. Next week I will probably be able to.

Reply:
No problem! I didn't really expect it to be honest, I just like to keep these things updated :)